### ISSUE TITLE
`unreachable_pub` false negative when an unreachable type is mentioned in a type alias

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust#116604]
I tried this code:

```rust
#![deny(unreachable_pub)]

mod foo {
    pub enum FooError {}
    pub type FooResult<T> = Result<T, FooError>;
}

pub use foo::FooResult;
```

I expected an `unreachable_pub` lint because `FooError` is unreachable. Instead, there is no diagnostic output.

The `FooError` type is in principle nameable via defining a suitable trait and associated type, but this is probably not the intent of the `unreachable_pub` lint; the author probably wants to be told to reexport `FooError`.

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-apple-darwin
release: 1.73.0
LLVM version: 17.0.2
```

```
rustc 1.75.0-nightly (2bbb61989 2023-10-04)
binary: rustc
commit-hash: 2bbb6198933abf193c0ef1d18d9d70f99eed59ba
commit-date: 2023-10-04
host: x86_64-apple-darwin
release: 1.75.0-nightly
LLVM version: 17.0.2
```

@rustbot label +A-lint +A-visibility

**Comment by Urgau** [comment]
I don't think this is a false positive, as your case is detected by the [`unnameable_types`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unnameable-types) lint which:
> detects types for which you can get objects of that type, but cannot name the type itself.

```text
error: enum `FooError` is reachable but cannot be named
 --> src/lib.rs:4:5
  |
4 |     pub enum FooError {}
  |     ^^^^^^^^^^^^^^^^^ reachable at visibility `pub`, but can only be named at visibility `pub(crate)`
  |
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(unnameable_types)]
  |         ^^^^^^^^^^^^^^^^
```

While the [`unreachable_pub`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unreachable-pub) does not lint on "leaked types":
> The unreachable_pub lint triggers for pub items not reachable from other crates - that means neither directly accessible, nor reexported, nor leaked through things like return types.

**Comment by kpreid** [comment]
That makes sense. Given this, I think it would be good if `unreachable_pub`’s documentation mentioned the `unnameable_types` lint. I will start a PR for that.
